### PR TITLE
PUBDEV-7452 add model reproducibility info to MOJO

### PR DESCRIPTION
**Pull request by koniecsveta**
https://0xdata.atlassian.net/projects/PUBDEV/issues/PUBDEV-7452

### CODE REVIEW

File: h2o-core/src/main/java/water/api/schemas3/ModelOutputSchemaV3.java [hunk 1]
```diff
@@ -49,6 +50,9 @@
   @API(help="Scoring history", direction=API.Direction.OUTPUT, level=API.Level.secondary)
   public TwoDimTableV3 scoring_history;
 
+  @API(help="Model reproducibility information", direction=API.Direction.OUTPUT, level=API.Level.secondary)
+  IcedHashMap<String, Object> reproducibility_information_map;
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/api/schemas3/ModelOutputSchemaV3.java:54]: should be public

File: h2o-core/src/main/java/water/HeartBeat.java [hunk 2]
```diff
@@ -48,6 +48,31 @@
   public long get_pojo_mem() { return _pojo_mem; }
   public long get_free_mem() { return _free_mem; }
   public long get_swap_mem() { return _swap_mem; }
+  
+  private long _total_mem;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/HeartBeat.java:52]: the HeartBeat structure should be kept minimal, most of this information is static and never changes. It is thus not a good idea to broadcast it with every single heartbeat.

File: h2o-core/src/main/java/water/api/schemas3/ModelOutputSchemaV3.java [hunk 3]
```diff
@@ -49,6 +50,9 @@
   @API(help="Scoring history", direction=API.Direction.OUTPUT, level=API.Level.secondary)
   public TwoDimTableV3 scoring_history;
 
+  @API(help="Model reproducibility information", direction=API.Direction.OUTPUT, level=API.Level.secondary)
+  public IcedHashMap<String, Object> reproducibility_information_map;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/api/schemas3/ModelOutputSchemaV3.java:54]: This should be TwoDimTableV3 instead

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java [hunk 4]
```diff
@@ -175,6 +175,39 @@ public SharedTreeOutput( SharedTree b) {
...
+              colHeaders.toArray(new String[0]),
+              colTypes.toArray(new String[0]),
+              colFormat.toArray(new String[0]),
+              "");
+
+      table.set(0, 0, "training_frame");
+      table.set(1, 0, "validation_frame");
+      table.set(2, 0, "calibration_frame");
+      table.set(0, 1, modelBuilder.train() != null ? modelBuilder.train().checksum() : -1);
+      table.set(1, 1, params._valid != null ? modelBuilder.valid().checksum() : -1);
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:202]: should this be modelBuilder.valid() != null? you can set the train/valid Frames directly

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java [hunk 5]
```diff
@@ -175,6 +175,39 @@ public SharedTreeOutput( SharedTree b) {
       _modelClassDist = _priorClassDist;
     }
 
+    @Override
+    public TwoDimTable createInputFramesInformationTable(ModelBuilder modelBuilder) {
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:179]: why is this defined on SharedTreeModel level?

the only tree-specific parameter is calibration frame

File: h2o-core/src/main/java/hex/Model.java [hunk 6]
```diff
@@ -504,6 +507,153 @@ public void addWarning(String s){
     Frame makeInteractions(Frame f);
   }
 
+  private static TwoDimTable createNodeInformationTable() {
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/Model.java:510]: Can we move this to a utility class? The model class is already large

File: h2o-core/src/main/java/water/H2ONode.java [hunk 7]
```diff
@@ -41,7 +39,8 @@
     
   transient private short _timestamp;
   transient private boolean _removed_from_cloud;
-  transient private volatile boolean _accessed_local_dkv; // Did this remote node ever accessed the local portion of DKV?   
+  transient private boolean _accessed_local_dkv; // Did this remote node ever accessed the local portion of DKV?
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/H2ONode.java:42]: the volatile modifier should be preserved, why was it removed?

File: h2o-core/src/main/java/water/H2ONode.java [hunk 8]
```diff
@@ -41,7 +39,8 @@
...
+  transient private IcedHashMap<String, Object> _configuration;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/H2ONode.java:43]: this is transient field - not serialized, what is the purpose of the _configuration field? it is not sent over the wire

File: h2o-core/src/main/java/water/api/CapabilitiesHandler.java [hunk 9]
```diff
@@ -18,7 +18,7 @@
     return entries;
   }
 
-  private ArrayList<CapabilityEntryV3> getRestAPIExtensionEntries(){
+  protected ArrayList<CapabilityEntryV3> getRestAPIExtensionEntries(){
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/api/CapabilitiesHandler.java:21]: why is this needed?

**Comment by michalkurka** [comment]
Look good but there are places where I have comments/concerns. Please let me know.

File: h2o-core/src/main/java/water/H2ONode.java (hunk 7 above)
Comment (koniecsveta) [reply h2o-core/src/main/java/water/H2ONode.java:42]: Hm this is weird, must have been done by a mistake. Will put back.

File: h2o-core/src/main/java/water/H2ONode.java (hunk 8 above)
Comment (koniecsveta) [reply h2o-core/src/main/java/water/H2ONode.java:43]: It keeps specific node configuration information in a hash map before they are written (along with some other information) to overall table Model#_reproducibility_information_table in createNodeInformationTable()

File: h2o-core/src/main/java/water/api/CapabilitiesHandler.java (hunk 9 above)
Comment (koniecsveta) [reply h2o-core/src/main/java/water/api/CapabilitiesHandler.java:21]: Will put back to private. I am using getRestApiExtensions() instead and must've forgotten to tidy this.

File: h2o-algos/src/main/java/hex/tree/SharedTreeModel.java (hunk 5 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/tree/SharedTreeModel.java:179]: will refactor so that it uses table = super.createInputFrames.. and just fills calibration frame here

**Comment by koniecsveta** [comment]
> Thank you @koniecsveta, congrats on your first MRTask ;)

🦄

File: h2o-core/src/main/java/hex/Model.java [hunk 10]
```diff
@@ -937,6 +949,46 @@ public void printTwoDimTables(StringBuilder sb, Object o) {
       printTwoDimTables(sb, this);
       return sb.toString();
     }
+
+    private TwoDimTable[] createReproducibilityInformationTable(ModelBuilder modelBuilder) {
+      TwoDimTable nodeInformation = ReproducibilityInformationUtils.createNodeInformationTable();
+      TwoDimTable clusterConfiguration = ReproducibilityInformationUtils.createClusterConfigurationTable();
+      TwoDimTable inputFramesInformation = createInputFramesInformationTable(modelBuilder);
+      TwoDimTable[] tables = {nodeInformation, clusterConfiguration, inputFramesInformation};
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/Model.java:957]: unnecessary variable

File: h2o-core/src/main/java/hex/Model.java [hunk 11]
```diff
@@ -937,6 +949,46 @@ public void printTwoDimTables(StringBuilder sb, Object o) {
...
+      return tables;
+    }
+
+    public TwoDimTable createInputFramesInformationTable(ModelBuilder modelBuilder) {
+      List<String> colHeaders = new ArrayList<>();
+      List<String> colTypes = new ArrayList<>();
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/Model.java:963]: this seems like it can be just an array, its never modified and the contents are the same, same for the other two Lists

File: h2o-core/src/main/java/hex/Model.java [hunk 12]
```diff
@@ -704,6 +707,7 @@ public void setNames(String[] names, String[] columntypes) {
     public long _end_time;
     public long _run_time;
     public long _total_run_time; // includes building of cv models
+    public static int INFORMATION_TABLE_NUM_ROWS = 2;
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/Model.java:710]: is this constant necessary? is it used somewhere else other than the getInformationTableNumRows method?

File: h2o-core/src/main/java/hex/Model.java [hunk 13]
```diff
@@ -937,6 +949,46 @@ public void printTwoDimTables(StringBuilder sb, Object o) {
...
+      table.set(0, 1, modelBuilder.train() != null ? modelBuilder.train().checksum() : -1);
+      table.set(1, 1, modelBuilder._valid != null ? modelBuilder.valid().checksum() : -1);
+      table.set(0, 2, modelBuilder.train() != null ? Arrays.toString(modelBuilder.train().anyVec().espc()) : -1);
+      table.set(1, 2, modelBuilder._valid != null ? Arrays.toString(modelBuilder.valid().anyVec().espc()) : -1);
+
+      return table;
+    }
+    
+    public int getInformationTableNumRows() {
+      return INFORMATION_TABLE_NUM_ROWS;
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/Model.java:990]: I would just put 2 here and a comment explaining what the two rows are

File: h2o-core/src/main/java/water/H2ONode.java [hunk 14]
```diff
@@ -3,12 +3,10 @@
...
 import water.nbhm.NonBlockingHashMapLong;
 import water.network.SocketChannelFactory;
-import water.util.ArrayUtils;
-import water.util.Log;
-import water.util.MathUtils;
-import water.util.UnsafeUtils;
+import water.util.*;
 
 import java.io.IOException;
+import java.lang.management.ManagementFactory;
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/H2ONode.java:9]: unused

File: h2o-core/src/main/java/water/HeartBeatThread.java [hunk 15]
```diff
@@ -118,7 +120,8 @@ public void run() {
         hb._process_total_ticks = -1;
         hb._process_num_open_fds = -1;
       }
-      hb._num_cpus = (short)Runtime.getRuntime().availableProcessors();
+      Runtime runtime = Runtime.getRuntime();
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/HeartBeatThread.java:123]: unnecessary var

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostOutput.java [hunk 16]
```diff
@@ -42,6 +44,21 @@ public XGBoostOutput(XGBoost b) {
...
+    TwoDimTable table = super.createInputFramesInformationTable(modelBuilder);
+    table.set(2, 0, "calibration_frame");
+    table.set(2, 1, params.getCalibrationFrame() != null ? params.getCalibrationFrame().checksum() : -1);
+    table.set(2, 2, params.getCalibrationFrame() != null ? Arrays.toString(params.getCalibrationFrame().anyVec().espc()) : -1);
+    return table;
+  }
+
+  @Override
+  public int getInformationTableNumRows() {
+    return super.getInformationTableNumRows() + 1;
```
Comment (honzasterba) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostOutput.java:59]: I woud add a comment about the +1 being for calibration frame

File: h2o-genmodel/src/main/java/hex/genmodel/ModelMojoReader.java [hunk 17]
```diff
@@ -200,6 +201,16 @@ private void readAll(final boolean readModelMetadata) throws IOException {
               .build();
       _model._modelAttributes = readModelSpecificAttributes();
     }
+    _model._reproducibilityInformation = readReproducibilityInformation() ;
+  }
+
+  protected Table[] readReproducibilityInformation() {
+    final JsonObject modelJson = ModelJsonReader.parseModelJson(_reader);
+    if (modelJson != null && modelJson.get("output") != null) {
+      Table[] tables = ModelJsonReader.readTableArray(modelJson, "output.reproducibility_information_table");
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/ModelMojoReader.java:210]: unnecessary var
